### DISCUSSION TITLE
Visualizing WSN cluster formation via heatmaps

### DISCUSSION THREAD

**Discussion by n-jay** [omnetpp/omnetpp#1093] [Show and tell]
### Introduction
While working on my research project on WSNs I needed a way to visualize the clustering patterns of group-based routing protocols especially when running large-scale simulations via the command line.

Since I couldn't find a way to directly extract the data using OMNET's built-in tools and wanted full control of the final output, I decided to develop a custom data extraction pipeline using **CSV** files and a **Python script**. 
I was inspired by the "heatmaps" shown in weather reports to showcase different phenomena such as wind and temperature fluctuations.

### Setup
I designed my protocol to write data to two CSV files at the conclusion the simulation; one consisting of _node location data_, and the second an _event log_.  

![Selection_159](https://github.com/omnetpp/omnetpp/assets/24616975/94423804-9204-4159-9e25-e11d7bcca010)

The event log was a chronological log of all packet transmission events. As observed in the image, it included metadata about the sender, recipient, type of packet, and the type of action.

Thereafter, I wrote a Python script to filter, aggregate and overlay data from the event log onto a scatterplot generated from the node location data. By leveraging the chronological arrangement of events in the log, I also developed a way to animate the "clustering behaviour" during the lifetime of the simulation.

### Result
The final output looked something like this:

https://github.com/omnetpp/omnetpp/assets/24616975/2ea0786c-85bf-4ad5-9929-5519c592c8d3

Later versions of the visualization were updated to overlay the node location scatterplot and the base-station position _over_ the heatmap for better context.

**Comment by rhornig** [comment]
WOww. Nice project. Congrats.

**Reply by n-jay** [reply]
Thanks Rudolf!
